File: lab5/haskell/diophantine.hs
import Text.Printf


extGcd :: Int -> Int -> (Int, Int, Int)
extGcd a 0 = (1, 0, a)
extGcd a b =
    let (x', y', gcdAB) = extGcd b (a `mod` b)
        (x, y) = (y', x' - y' * (a `div` b))
    in (x, y, gcdAB)

de :: (Int, Int) -> (Int, Int, Int)
de (a, b) = extGcd a b


main :: IO ()
main = do
    let (a, b) = (30, 42)
        (x, y, z) = de (a, b)

    printf "Solution to the equation %dx + %dy = gcd(%d, %d):\n" a b a b
    printf "x = %d | y = %d | gcd(%d, %d) = %d\n" x y a b z
    printf "Check: %d * %d + %d * %d = %d\n" a x b y (a * x + b * y)

File: lab5/prolog/diophantine.pl
extGcd(A, 0, 1, 0, A).
extGcd(A, B, X, Y, Z) :-
    B > 0,
    Tmp is A mod B,
    extGcd(B, Tmp, X1, Y1, Z),
    X is Y1,
    Y is X1 - Y1 * (A // B).

de(A, B, X, Y, Z) :-
    extGcd(A, B, X, Y, Z).


main :-
    A = 30,
    B = 42,
    de(A, B, X, Y, Z),
    format("Solution to the equation ~dx + ~dy = gcd(~d, ~d):~n", [A, B, A, B]),
    format("x = ~d | y = ~d | gcd(~d, ~d) = ~d~n", [X, Y, A, B, Z]),
    format("Check: ~d * ~d + ~d * ~d = ~d~n", [A, X, B, Y, A * X + B * Y]).

:- main.
